Initial implementation of PathStatus and EndPathStatus products

wddgit opened the pull request:
The producers for these products are run automatically after
a Path or EndPath completes. They will be useful in a
multithreaded context as they allow creating a dependence
between a module and a path in a way similar to how
a consumes declaration creates a dependence between
two modules.

wddgit committed:
Initial implementation of PathStatus and EndPathStatus products

The producers for these products are run automatically after
a Path or EndPath completes. They will be useful in a
multithreaded context as they allow creating a dependence
between a module and a path in a way similar to how
a consumes declaration creates a dependence between
two modules.

cmsbuild commented:
A new Pull Request was created by  @wddgit (W. David Dagenhart) for master.

It involves the following packages:

DataFormats/Common
DataFormats/Provenance
FWCore/Concurrency
FWCore/Framework
FWCore/Integration
FWCore/Modules
IOMC/RandomEngine


@cmsbuild, @smuzaffar, @Dr15Jones, @davidlange6 can you please review it and eventually sign? Thanks.
@makortel, @Martin-Grunewald, @ebrondol this is something you requested to watch as well.
@davidlange6 you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

wddgit commented:
I expect Chris will want to review this one before it gets merged and he is not back from vacation until 12 June. I submitted it anyway because I want to see if the PR tests find any problems. As far as I know there is no need for this PR before the 12th.

wddgit commented:
please test

cmsbuild commented:
-1

Tested at: 967bf8aa185d89d7df90c69aba47e1ed8ad54dc6

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-19052/20272/summary.html

I found follow errors while testing this PR

**Failed tests:  UnitTests**

* **Unit Tests**:

I found errors in the following unit tests: 
 
 ---> test testRecoMETMETProducers had ERRORS

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-19052/20272/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 1697 differences found in the comparisons
 - DQMHistoTests: Total files compared: 23
 - DQMHistoTests: Total histograms compared: 1786106
 - DQMHistoTests: Total failures: 38947
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 1746986
 - DQMHistoTests: Total skipped: 173
 - DQMHistoTests: Total Missing objects: 0
 - Checked 94 log files, 14 edm output root files, 23 DQM output files

wddgit commented:
The unit test failure is not related to this PR.

Dr15Jones commented:
Please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
-1

Tested at: 967bf8aa185d89d7df90c69aba47e1ed8ad54dc6

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-19052/20513/summary.html

I found follow errors while testing this PR

**Failed tests:  UnitTests**

* **Unit Tests**:

I found errors in the following unit tests:

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-19052/20513/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 22
 - DQMHistoTests: Total histograms compared: 1817858
 - DQMHistoTests: Total failures: 19835
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 1797857
 - DQMHistoTests: Total skipped: 166
 - DQMHistoTests: Total Missing objects: 0
 - Checked 90 log files, 14 edm output root files, 22 DQM output files

Dr15Jones commented on DataFormats/Common/src/classes_def.xml:
```diff
@@ -115,7 +115,11 @@
  </class>
- <class name="edm::Wrapper<edm::HLTPathStatus>" splitLevel="0"/>
+ <class name="edm::Wrapper<edm::HLTPathStatus>"  persistent="false" splitLevel="0"/>
+ <class name="edm::EndPathStatus" ClassVersion="10">
```
ClassVersion for new classes should start at 3.

Dr15Jones commented on FWCore/Framework/src/GlobalSchedule.cc:
```diff
@@ -17,6 +19,8 @@
 
 namespace edm {
   GlobalSchedule::GlobalSchedule(std::shared_ptr<TriggerResultInserter> inserter,
+                                 std::vector<edm::propagate_const<std::shared_ptr<PathStatusInserter>>>& pathStatusInserters,
```
Why pass these to the GlobalSchedule? Paths are not run in global transitions.

Dr15Jones commented on FWCore/Framework/src/Path.h:
```diff
@@ -117,8 +122,9 @@ namespace edm {
     WaitingTaskList waitingTasks_;
     std::atomic<bool>* stopProcessingEvent_;
 
+    PathStatusInserter* pathStatusInserter_;
```
Can this be `PathStatusInserter const*` ?

Dr15Jones replied:
I see it can't be.

Dr15Jones commented on FWCore/Framework/src/PathStatusInserter.cc:
```diff
@@ -0,0 +1,27 @@
+  void
+  PathStatusInserter::setPathStatus(StreamID const& streamID,
+                                    HLTPathStatus const& hltPathStatus) {
+    hltPathStatus_[streamID.value()] = hltPathStatus;
```
I don't think this is thread safe. We need something to force a synchronization across threads. Let's talk about this.

Dr15Jones replied:
After discussion, we determined this is thread safe since the action of setting a value is done on the same thread just before the value will be read.

Dr15Jones commented on FWCore/Framework/src/PathsAndConsumesOfModules.cc:
```diff
@@ -141,9 +141,13 @@ namespace edm {
     std::unordered_map<unsigned int, std::string> moduleIndexToNames;
-    
+
+    std::map<std::string, unsigned int> pathStatusInserterModuleLabelToModuleID;
```
How about switching to a `std::unordered_map`? They tend to be more memory efficient.

Dr15Jones commented on FWCore/Framework/src/StreamSchedule.cc:
```diff
@@ -539,17 +535,35 @@ namespace edm {
-                                            EventSetup const& es)
-  {
+                                            EventSetup const& es,
+                                            std::vector<edm::propagate_const<std::shared_ptr<PathStatusInserter>>>& pathStatusInserters) {
```
Maybe it would be better to determine the empty path inserters up front and just keep that list? That way there are no filtering to do during this call.

Dr15Jones commented on FWCore/Framework/src/Worker.h:
```diff
@@ -104,6 +105,13 @@ namespace edm {
                                   typename T::Context const* context);
 
+    template <typename T>
+    void runStatusInserter(typename T::MyPrincipal const& ep,
```
I'm not a fan of extending the Worker this way. What is is needed for?

wddgit committed:
Improvements in response to comments from Chris

cmsbuild commented:
Pull request #19052 was updated. @cmsbuild, @smuzaffar, @Dr15Jones, @davidlange6 can you please check and sign again.

wddgit commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

Dr15Jones commented on FWCore/Framework/src/StreamSchedule.cc:
```diff
@@ -544,26 +544,27 @@ namespace edm {
-        );
+    HLTPathStatus hltPathStatus(hlt::Pass, 0);
+    for (int empty_trig_path : empty_trig_paths_) {
+      results_->at(empty_trig_path) = hltPathStatus;
```
Do we still need `at` here?

wddgit replied:
In the class HLTGlobalStatus both the operator[] and the at function call the at function of the vector contained in the class. Calling at seemed more honest. I suppose I could change HLTGlobalStatus if you want me to.
http://cmslxr.fnal.gov/source/DataFormats/Common/interface/HLTGlobalStatus.h#0058

wddgit commented:
The last commit I pushed included all the fixes related to Chris's comments. Here
is a summary with responses to the comments and what I changed in the
new commit.

1. ClassVersion for new classes should start at 3.

    Done

2. Why pass these to the GlobalSchedule? Paths are not run in global transitions.

    The AllWorkers list in GlobalSchedule is used for a few things like
    printing the module timing summary at the end of job and the function
    getAllModuleDescriptions, the function replaceModule, used in
    PathsAndConsumesOfModules.cc and maybe some other places ...

3. I don't think this is thread safe. We need something to force a synchronization
across threads. Let's talk about this.

    After discussing this and looking at it more closely we decided
    this was thread safe.

4. How about switching to a std::unordered_map? They tend to be more memory efficient.

    Done

5. Maybe it would be better to determine the empty path inserters up front and just
keep that list? That way there are no filtering to do during this call.

    We discussed this and decided it was OK as is.

6. I'm not a fan of extending the Worker this way. What is is needed for?

    We discussed this code and I modified it and removed the runStatusInserter
    function. I think I fixed this in the way we discussed.

cmsbuild commented:
-1

Tested at: 8d88761874a61dbf6535d588268c2eb9b1f46965

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-19052/20652/summary.html

I found follow errors while testing this PR

**Failed tests:  UnitTests**

* **Unit Tests**:

I found errors in the following unit tests: 
 
 ---> test testTauEmbeddingProducers had ERRORS

cmsbuild commented:
Comparison job queued.

wddgit commented:
The unit test failure is unrelated to this PR. It also occurs in IBs without these changes.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-19052/20652/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 22
 - DQMHistoTests: Total histograms compared: 1803204
 - DQMHistoTests: Total failures: 29458
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 1773580
 - DQMHistoTests: Total skipped: 166
 - DQMHistoTests: Total Missing objects: 0
 - Checked 90 log files, 14 edm output root files, 22 DQM output files

Dr15Jones commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
-1

Tested at: 8d88761874a61dbf6535d588268c2eb9b1f46965

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-19052/20766/summary.html

I found follow errors while testing this PR

**Failed tests:  UnitTests**

* **Unit Tests**:

I found errors in the following unit tests:

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-19052/20766/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 22
 - DQMHistoTests: Total histograms compared: 1803136
 - DQMHistoTests: Total failures: 29506
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 1773464
 - DQMHistoTests: Total skipped: 166
 - DQMHistoTests: Total Missing objects: 0
 - Checked 90 log files, 14 edm output root files, 22 DQM output files

wddgit commented:
@smuzaffar Twice on this PR (the last time and a couple times before that), there has been a message that the unit tests failed, but no unit test failures are listed. And I searched in the log and cannot find any occurrences of the string ERRORS in the log file. Do you have any idea what could cause that?

smuzaffar commented:
@wddgit , it looks like the unit tests jenkins job was timed out after 8h26mins (as this PR nearly build full cmssw and ran all unit tests). 
```
>> Tests for package DQMServices/Components ran.
gmake: *** wait: No child processes.  Stop.
gmake: *** Waiting for unfinished jobs....
gmake: *** wait: No child processes.  Stop.
```

davidlange6 commented:
hi @wddgit  - could you rebase this PR. Thanks

cmsbuild commented:
Pull request #19052 was updated. @cmsbuild, @smuzaffar, @Dr15Jones, @davidlange6 can you please check and sign again.

wddgit commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

Dr15Jones commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs after it passes the integration tests. This pull request requires discussion in the ORP meeting before it's merged. @davidlange6, @smuzaffar

cmsbuild commented:
-1

Tested at: c80e619cbb40eb7721c756fa9836c5079e7659d1


The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/a44f5ad0863b067cbc88f463b075504044dea5db
https://github.com/cms-sw/cmssw/commit/0128241156fdface391f39e4a6b21de18508b2ba
https://github.com/cms-sw/cmssw/commit/a4629aa8ab04db10489542f4da9d2adb3d0b6208
https://github.com/cms-sw/cmssw/commit/98ae32ff78c423a0e09515b5407b7afebb8a7e4e
https://github.com/cms-sw/cmssw/commit/d18eb61dd685a73fee721f1dcd3457da5015e75e
https://github.com/cms-sw/cmssw/commit/fc51b5be13c935a544fbc6d7dc42b47194380ab5
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-19052/20922/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-19052/20922/git-merge-result

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-19052/20922/summary.html

I found follow errors while testing this PR

**Failed tests:  UnitTests**

* **Unit Tests**:

I found errors in the following unit tests: 
 
 ---> test testRecoMETMETProducers had ERRORS
---> test runtestRecoEgammaPhotonIdentification had ERRORS
---> test runtestRecoEgammaElectronIdentification had ERRORS



The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/a44f5ad0863b067cbc88f463b075504044dea5db
https://github.com/cms-sw/cmssw/commit/0128241156fdface391f39e4a6b21de18508b2ba
https://github.com/cms-sw/cmssw/commit/a4629aa8ab04db10489542f4da9d2adb3d0b6208
https://github.com/cms-sw/cmssw/commit/98ae32ff78c423a0e09515b5407b7afebb8a7e4e
https://github.com/cms-sw/cmssw/commit/d18eb61dd685a73fee721f1dcd3457da5015e75e
https://github.com/cms-sw/cmssw/commit/fc51b5be13c935a544fbc6d7dc42b47194380ab5
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-19052/20922/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-19052/20922/git-merge-result

cmsbuild commented:
Comparison job queued.

wddgit commented:
Test failures are not related to this PR.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-19052/20922/summary.html

Comparison Summary:
 - You potentially added 3649 lines to the logs
 - ROOTFileChecks: Some differences in event products or their sizes found
 - Reco comparison results: 1702 differences found in the comparisons
 - DQMHistoTests: Total files compared: 22
 - DQMHistoTests: Total histograms compared: 1690801
 - DQMHistoTests: Total failures: 48319
 - DQMHistoTests: Total nulls: 34
 - DQMHistoTests: Total successes: 1642282
 - DQMHistoTests: Total skipped: 166
 - DQMHistoTests: Total Missing objects: 0
 - Checked 90 log files, 14 edm output root files, 22 DQM output files

davidlange6 commented:
merge

smuzaffar commented:
@wddgit  there are now merge conflicts b/w master and CMSSW_9_3_DEVEL_X IBs. Can you please resolve the conflicts or let me know if we want to keep the CMSSW_9_3_DEVEL_X  changes or force override changes from master for FWCore/Integration/test/unit_test_outputs/testSubProcess.grep2.txt

Dr15Jones commented:
@smuzaffar the two changes are now in master so it is OK to override.

smuzaffar commented:
done @Dr15Jones
